google/cloud/storage/tests/object_media_integration_test.cc

```diff
@@ -317,6 +317,123 @@ TEST_F(ObjectMediaIntegrationTest, ReadMixedChunks) {
   EXPECT_STATUS_OK(status);
 }
 
+/// @test Read the last chunk of an object.
+TEST_F(ObjectMediaIntegrationTest, ReadLastChunk) {
+  StatusOr<Client> client = Client::CreateDefaultClient();
+  ASSERT_STATUS_OK(client);
+
+  std::string bucket_name = flag_bucket_name;
+  auto object_name = MakeRandomObjectName();
+
+  // This produces an object larger than 3MiB, but with a size that is not a
+  // multiple of 128KiB.
+  auto const kKiB = 1024L;
+  auto const kMiB = 1024L * kKiB;
+  auto const object_size = 3 * kMiB + 129 * kKiB;
+  auto const lines = object_size / 128;
+  std::string large_text;
+  for (long i = 0; i != lines; ++i) {
+    auto line = google::cloud::internal::Sample(generator_, 127,
+                                                "abcdefghijklmnopqrstuvwxyz"
+                                                "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                                                "012456789");
+    large_text += line + "\n";
+  }
+  static_assert(object_size % 128 == 0, "Object must be multiple of line size");
```

Comment:
Doesn't this require that object_size is (explicitly) constexpr?

Reply:
I do not think it *requires* it. `object_size` is a constant expression [1], its value is known at compile time, regardless of whether it is declared `constexpr` or not.  This is how we did `const int n = 10; int array[n];` in C++98 and C++03.  But I think it is probably more readable if we *say* that this is indeed a `constexpr`, and therefore changed the code.  I think I will adopt this as a practice.


[1]: I am, of course, referring to the expression that evaluates `object_size`, which is not the same thing as the object `object_size`, which is an object of type `long const` (or thereabouts, I think at least one of the literals has an `L` and propagates to all the constants). But this is me trying to be pedantic, which is bound to blow up in my face, I am terrible at these things :grinning:

Reply:
Ack.